File: webhook-audit-blotter/dotnet/src/WebhookAuditBlotter/WebhookHandler.cs
using System.Text;
using System.Text.Json;

namespace WebhookAuditBlotter;

public enum WebhookOutcomeStatus
{
    Accepted,
    BadSignature,
    BadRequest,
}

public sealed record WebhookOutcome(WebhookOutcomeStatus Status, long? RowId = null, string? Reason = null);

public sealed record IncomingDelivery(
    byte[] RawBody,
    string? SignatureHeader,
    string? EventTypeHeader);

public interface IAppendOnlyStore
{
    long Append(AppendArgs args);
}

public interface IClock
{
    string NowIso();
}

public sealed class SystemClock : IClock
{
    public string NowIso() => DateTimeOffset.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffZ", System.Globalization.CultureInfo.InvariantCulture);
}

/// <summary>
/// Framework-agnostic webhook orchestration: verify signature,
/// persist, produce an outcome.
/// </summary>
public sealed class WebhookHandler
{
    private readonly string secret;
    private readonly IAppendOnlyStore store;
    private readonly IClock clock;

    public WebhookHandler(string secret, IAppendOnlyStore store, IClock clock)
    {
        this.secret = secret;
        this.store = store;
        this.clock = clock;
    }

    public WebhookOutcome Handle(IncomingDelivery delivery)
    {
        if (!SignatureVerifier.Verify(delivery.RawBody, delivery.SignatureHeader, this.secret))
        {
            return new WebhookOutcome(WebhookOutcomeStatus.BadSignature);
        }
        var bodyText = Encoding.UTF8.GetString(delivery.RawBody);
        if (!IsJsonObject(bodyText))
        {
            return new WebhookOutcome(WebhookOutcomeStatus.BadRequest, Reason: "Body is not a JSON object.");
        }
        var eventType = (delivery.EventTypeHeader ?? string.Empty).Trim();
        if (eventType.Length == 0)
        {
            return new WebhookOutcome(WebhookOutcomeStatus.BadRequest, Reason: "Missing X-Evagene-Event header.");
        }
        var id = this.store.Append(new AppendArgs(this.clock.NowIso(), eventType, bodyText));
        return new WebhookOutcome(WebhookOutcomeStatus.Accepted, RowId: id);
    }

    private static bool IsJsonObject(string text)
    {
        try
        {
            using var doc = JsonDocument.Parse(text);
            return doc.RootElement.ValueKind == JsonValueKind.Object;
        }
        catch (JsonException)
        {
            return false;
        }
    }
}

File: webhook-audit-blotter/node/src/webhookHandler.ts
/**
 * Framework-agnostic webhook orchestration: verify signature, persist,
 * produce an outcome.  Server adapters translate the outcome into an
 * HTTP response.
 */

import type { AppendArgs } from './eventStore.js';
import { verifySignature } from './signatureVerifier.js';

export type WebhookOutcome =
  | { readonly status: 'accepted'; readonly rowId: number }
  | { readonly status: 'bad_signature' }
  | { readonly status: 'bad_request'; readonly reason: string };

export interface AppendOnlyStore {
  append(args: AppendArgs): number;
}

export interface Clock {
  nowIso(): string;
}

export interface WebhookHandlerOptions {
  readonly secret: string;
  readonly store: AppendOnlyStore;
  readonly clock: Clock;
}

export interface IncomingDelivery {
  readonly rawBody: Buffer;
  readonly signatureHeader: string | undefined;
  readonly eventTypeHeader: string | undefined;
}

export class WebhookHandler {
  constructor(private readonly options: WebhookHandlerOptions) {}

  handle(delivery: IncomingDelivery): WebhookOutcome {
    if (!verifySignature(delivery.rawBody, delivery.signatureHeader, this.options.secret)) {
      return { status: 'bad_signature' };
    }
    const bodyText = delivery.rawBody.toString('utf8');
    if (!isJsonObject(bodyText)) {
      return { status: 'bad_request', reason: 'Body is not a JSON object.' };
    }
    const eventType = (delivery.eventTypeHeader ?? '').trim();
    if (!eventType) {
      return { status: 'bad_request', reason: 'Missing X-Evagene-Event header.' };
    }
    const rowId = this.options.store.append({
      receivedAt: this.options.clock.nowIso(),
      eventType,
      body: bodyText,
    });
    return { status: 'accepted', rowId };
  }
}

function isJsonObject(text: string): boolean {
  try {
    const parsed: unknown = JSON.parse(text);
    return typeof parsed === 'object' && parsed !== null && !Array.isArray(parsed);
  } catch {
    return false;
  }
}
